src/main/java/com/gtnewhorizons/gtnhgradle/modules/JVMDowngraderModule.java

```diff
@@ -0,0 +1,595 @@
+        final TaskProvider<? extends AbstractArchiveTask> publishableDevJar = configureJarTaskChain(
+            gtnh,
+            project,
+            downgradeTarget,
+            targetVersion,
+            shadeStubs);
+        configureRunTasks(gtnh, project, downgradeTasks, publishableDevJar);
+        validateLombokVersion(gtnh, project);
+    }
+
+    private void applyJvmdgPlugin(Project project, GTNHGradlePlugin.GTNHExtension gtnh, JavaVersion targetVersion) {
+        project.getPluginManager()
+            .apply("xyz.wagyourtail.jvmdowngrader");
+
+        final JVMDowngraderExtension jvmdgExt = project.getExtensions()
+            .getByType(JVMDowngraderExtension.class);
+        jvmdgExt.getDowngradeTo()
+            .set(targetVersion);
+
+        configureMultiReleaseJar(project, jvmdgExt, gtnh);
+    }
+
+    private void configureMultiReleaseJar(Project project, JVMDowngraderExtension jvmdgExt,
+        GTNHGradlePlugin.GTNHExtension gtnh) {
+        final int toolchainVersion = gtnh.getEffectiveToolchainVersion()
+            .get();
+        final Set<Integer> multiReleaseVersions = gtnh.getJvmDowngraderMultiReleaseVersions()
+            .get();
+
+        if (multiReleaseVersions.isEmpty()) {
+            return;
+        }
+
+        final Set<JavaVersion> intermediateVersions = new HashSet<>();
+        for (int version : multiReleaseVersions) {
+            if (version == toolchainVersion) {
+                jvmdgExt.getMultiReleaseOriginal()
+                    .set(true);
+            } else if (version < toolchainVersion) {
+                intermediateVersions.add(JavaVersion.toVersion(version));
+            } else {
+                project.getLogger()
+                    .warn(
+                        "jvmDowngraderMultiReleaseVersions contains {} which exceeds toolchain version {}. "
+                            + "Ignoring - cannot generate bytecode for a higher Java version than the compiler.",
+                        version,
+                        toolchainVersion);
+            }
+        }
+
+        if (!intermediateVersions.isEmpty()) {
+            jvmdgExt.getMultiReleaseVersions()
+                .set(intermediateVersions);
+        }
+    }
+
+    private void configureCompileDependencies(Project project, GTNHGradlePlugin.GTNHExtension gtnh, int downgradeTarget,
+        JvmDowngraderStubsProvider stubsProvider) {
+        final DependencyHandler deps = project.getDependencies();
+        final String downgradedApiDep = UpdateableConstants.jvmdgApiDependency(downgradeTarget);
+        deps.add(JavaPlugin.COMPILE_ONLY_CONFIGURATION_NAME, downgradedApiDep);
+        deps.add(JavaPlugin.TEST_RUNTIME_ONLY_CONFIGURATION_NAME, downgradedApiDep);
+
+        addJabelStub(project, deps);
+
+        if (!stubsProvider.shouldShadeStubs() && !stubsProvider.isExternal()) {
+            deps.getConstraints()
+                .add(
+                    JavaPlugin.IMPLEMENTATION_CONFIGURATION_NAME,
```

Comment:
This is not enough and doesn't catch compileOnlyApi etc. dependencies, see how it's handled for unimixins